File: Arrays/AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        index=0
        newString=0
        while l1:
            newString+=((10**index)*l1.val)
            index+=1
            l1=l1.next
        index=0
        while l2:
            newString+=((10**index)*l2.val)
            index+=1
            l2=l2.next
        val =newString
        node=ListNode(int(str(val)[len(str(val))-1]))
        test=node
        for i in range(len(str(val))-2,-1,-1):
            node.next=ListNode(int(str(val)[i]))
            node=node.next
        
        return test

File: Arrays/AddTwoNumbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include <stdlib.h>

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    int index = 0;
    long long new_string = 0;

    (void)index;
    (void)new_string;

    /* Perform digit-by-digit addition to preserve exact functionality for arbitrarily large numbers. */
    int carry = 0;

    struct ListNode testNode;
    testNode.val = 0;
    testNode.next = NULL;

    struct ListNode *tail = &testNode;

    while (l1 != NULL || l2 != NULL || carry != 0) {
        int sum = carry;

        if (l1 != NULL) {
            sum += l1->val;
            l1 = l1->next;
        }
        if (l2 != NULL) {
            sum += l2->val;
            l2 = l2->next;
        }

        carry = sum / 10;

        struct ListNode *node = (struct ListNode *)malloc(sizeof(struct ListNode));
        if (node == NULL) {
            return NULL;
        }
        node->val = sum % 10;
        node->next = NULL;

        tail->next = node;
        tail = node;
    }

    return testNode.next;
}
